internal/applier/helm_test.go

```diff
@@ -0,0 +1,315 @@
+		testDesiredRelease := *testCurrentRelease
+		testDesiredRelease.Manifest = "do-not-match-current"
+
+		mockAcg := &mockActionGetter{
+			upgradeErr:      errors.New("failed upgrading chart"),
+			dryRunUpgradeOK: true, // let the dry-run upgrade pass
+			currentRel:      testCurrentRelease,
+			desiredRel:      &testDesiredRelease,
+		}
+		mockPf := &mockPreflight{upgradeErr: errors.New("failed during upgrade pre-flight check")}
+		helmApplier := applier.Helm{ActionClientGetter: mockAcg, Preflights: []applier.Preflight{mockPf}}
+
+		objs, state, err := helmApplier.Apply(context.TODO(), validFS, testCE, testObjectLabels, testStorageLabels)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "upgrade pre-flight check")
+		require.Equal(t, applier.StateNeedsUpgrade, state)
+		require.Nil(t, objs)
+	})
+
+	t.Run("fails during upgrade", func(t *testing.T) {
+		testDesiredRelease := *testCurrentRelease
+		testDesiredRelease.Manifest = "do-not-match-current"
+
+		mockAcg := &mockActionGetter{
+			upgradeErr:      errors.New("failed upgrading chart"),
+			dryRunUpgradeOK: true, // let the dry-run upgrade pass
+			currentRel:      testCurrentRelease,
+			desiredRel:      &testDesiredRelease,
+		}
+		mockPf := &mockPreflight{}
+		helmApplier := applier.Helm{ActionClientGetter: mockAcg, Preflights: []applier.Preflight{mockPf}}
+
+		objs, state, err := helmApplier.Apply(context.TODO(), validFS, testCE, testObjectLabels, testStorageLabels)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "upgrading chart")
+		require.Equal(t, applier.StateNeedsUpgrade, state)
+		require.Nil(t, objs)
+	})
+
+	t.Run("fails during upgrade reconcile (StateUnchanged)", func(t *testing.T) {
+		// make sure desired and current are the same this time
+		testDesiredRelease := *testCurrentRelease
+
+		mockAcg := &mockActionGetter{
+			reconcileErr:    errors.New("failed reconciling charts"),
+			dryRunUpgradeOK: true, // let the dry-run upgrade pass
+			currentRel:      testCurrentRelease,
+			desiredRel:      &testDesiredRelease,
+		}
+		mockPf := &mockPreflight{}
+		helmApplier := applier.Helm{ActionClientGetter: mockAcg, Preflights: []applier.Preflight{mockPf}}
+
+		objs, state, err := helmApplier.Apply(context.TODO(), validFS, testCE, testObjectLabels, testStorageLabels)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "reconciling charts")
+		require.Equal(t, applier.StateUnchanged, state)
+		require.Nil(t, objs)
+	})
+
+	t.Run("successful upgrade", func(t *testing.T) {
+		testDesiredRelease := *testCurrentRelease
+		testDesiredRelease.Manifest = validManifest
+
+		mockAcg := &mockActionGetter{
+			currentRel: testCurrentRelease,
+			desiredRel: &testDesiredRelease,
+		}
+		helmApplier := applier.Helm{ActionClientGetter: mockAcg}
+
+		objs, state, err := helmApplier.Apply(context.TODO(), validFS, testCE, testObjectLabels, testStorageLabels)
+		require.NoError(t, err)
+		require.Equal(t, applier.StateNeedsUpgrade, state)
```

Comment:
Is that the correct state expected when the upgrade occurs with success?

Reply:
please see https://github.com/operator-framework/operator-controller/pull/1539#discussion_r1906034448

Reply:
@joelanford @bentito 
Same here. Is it not a bug? 
Should we have the state `StateNeedsUpgrade ` when the upgrade was successful?

Reply:
so just my 2c after looking at the code - `Apply` is only ever called by the cluster extension controller and the status it returns is currently being ignored altogether:

> managedObjs, _, err := r.Applier.Apply(ctx, unpackResult.Bundle, ext, objLbls, storeLbls)

In case there was no error, the extension status will be set as installed in status conditions, so there is no differentiation between reconcilier having to internally upgrade an existing release or just do a fresh install.
IMHO, that probably makes sense as the final status info to pass (current bundle status being 'installed' rather than 'upgraded').

Reply:
Yes, I think we do not update it after that,
It is out of the scope anyway.

I will put it in my TODO list to do a few tests and see how is the UX 
If I think that is confusing, then I open an issue for us to discuss. 

Thank you a lot. Marking as solved